Code/Tools/Libs/ToolsFoundation/Command/Implementation/TreeCommands.cpp

```diff
@@ -702,7 +702,7 @@ ezStatus ezSetObjectPropertyCommand::DoInternal(bool bRedo)
 
   if (!bRedo)
   {
-    EZ_ASSERT_DEBUG(m_NewValue.GetType() != ezVariantType::StringView && m_NewValue.GetType() != ezVariantType::TypedPointer, "Variants that are stored in the command history must hold ownership of their value.");
```

Comment:
This assert would often fire (now) when editing arrays that contain resource references (string views).
Looks like this assert isn't needed? Nothing breaks without it.

Reply:
So this assert makes sure that the undo history does not have pointers to external memory which could be freed and cause crashes later. In this specific case it only works because the stringview points to the RTTI default value property so it is static. Unless it's a phantom type in which case updating type info would invalidate the default property and thus invalidate this memory address. So the code should not be changed.
Instead, go up the callstack on the assert and look for a good location to fix the type of the change.

One option would be to update `ezDocumentObjectConverterReader::ApplyDiff` to call:
```C++
      if (pProp->GetFlags().IsAnySet(ezPropertyFlags::IsEnum | ezPropertyFlags::Bitflags) || bIsValueType)
      {
        const ezVariantType::Enum memberType = ezToolsReflectionUtils::GetStorageType(pProp);
        if (bIsValueType && op.m_Value.GetType() != memberType)
          op.m_Value = op.m_Value.ConvertTo(memberType);

        pObjectAccessor->SetValue(pObject, pProp, op.m_Value).IgnoreResult();
      }
```
`ezToolsReflectionUtils::GetStorageType` returns a value that can be stored in the object as well as in the undo history. There might be a better place to update the type, for example when the default value is computed, but I'd have to have a deeper look at the code. One example would be to replace all calls in tools code to `ezReflectionUtils::GetDefaultValue` with `ezToolsReflectionUtils::GetStorageDefault` which does the same but takes care of the storage type. However, the later one does not support given a sub-index to the function so some more work is required for that kind of fix.

Reply:
fixed